### ISSUE TITLE
data-controlled ODE with general control signal

### ISSUE THREAD

**Issue by Wang-Tianyu** [DiffEqML/torchdyn#121]
Hi, thanks for the great library.

I am interested in the ODE of the form dz(s)/ds = f_theta (s, r, z(s)) where r can be a vector independent of input x.
Does the current implementation support this feature?

**Comment by Zymrael** [comment]
Hey, thanks for using `torchdyn`.

What you ask is not currently supported with the `DataControl` layer, but can be done pretty easily in two ways:
* You can have your vector field `nn.Module` save `r` in `self.r` and use it in a forward that calls `(s, z)` only.
* [This](https://github.com/DiffEqML/torchdyn/blob/1ee57d7b6d9b9991234c5b96780c01002f1ea712/torchdyn/core/neuralde.py#L86) is where the data control is set by default. The only module with `u` we consider is [DataControl](https://github.com/DiffEqML/torchdyn/blob/1ee57d7b6d9b9991234c5b96780c01002f1ea712/torchdyn/nn/node_layers.py#L74), which triggers the above to assign the current input to `module.u`. Depending on how general you'd prefer your implementation to be, you can simply modify `DataControl`

```
class CustomDataControl(nn.Module):
    """Data-control module. Allows for datasets-control inputs at arbitrary points of the DEFunc
    """
    def __init__(self, r):
        super().__init__()
        self.u = None
        self.r = r

    def forward(self, x):
        return torch.cat([x, self.r], 1).to(x)
```
which is slightly hacky but works, or you can alter the logic in `_prep_integration` to allow for custom assignments to `module.u`.

**Comment by Wang-Tianyu** [comment]
Hey, thanks for your quick reply.

I managed to make it work following your suggestion!

One remaining question is, in this case, does the general condition r get any gradient?
I am interested in building conditional CNF like this work https://github.com/stevenygd/PointFlow
Do you think that is doable with this library?

**Comment by Zymrael** [comment]
Glad to hear it worked. Depending on your implementation, it should work just fine and get gradients. It is certainly possible to build PointFlows with `torchdyn` :)

**Comment by alokwarey** [comment]
I am trying to do something similar. However, my control inputs vary with each batch in the forward pass. My code snippet below. Will passing next(iter(controlloader)) to the CustomDataControl ensure that the right control inputs go with the corresponding batch?

u = torch.Tensor(u_train_scaled)
controlloader = data.DataLoader(u, batch_size=data_sample, shuffle=False)
it = iter(controlloader)

class CustomDataControl(nn.Module):
    """Data-control module. Allows for datasets-control inputs at arbitrary points of the DEFunc
    """
    def __init__(self,r):
        super().__init__()
        self.u = None
        self.r = r

    def forward(self, x):        
        return torch.cat([x, self.r], 1).to(x)

f = nn.Sequential(CustomDataControl(next(it)),
                  nn.Linear(num_inputs+num_control, 64),
                  nn.Tanh(),
                  nn.Linear(64, num_outputs))
